### ISSUE TITLE
[glsl-in] Consider supporting combined image/samplers

### ISSUE THREAD

**Issue by magcius** [gfx-rs/wgpu#4342]
My use case for Naga is converting from GLSL ES to WGSL at runtime. GLSL ES only supports combined image/samplers, and it would be quite annoying for me to change my code to do something else. Many typical approaches to this problem are limited by the restrictions of GLSL.

For instance, my current policy is to have texture bindings and samplers take consecutive bindings.

I would be fine if the GLSL is parsed into an IR that can't be output by the WGSL backend, and I would have to convert it to the proper policy myself.

**Comment by kent-h** [comment]
Hacked together a solution for myself, posting here in case anyone else finds it helpful:
https://github.com/kent-h/spirv-image-sampler-combiner

TL;DR: Made a tool that rewrites SPIR-V files with separate `Image` and `Sampler` bindings into a combined `ImageSampler`.  I'm sure there are edge cases that don't work yet, and please be aware that it doesn't currently handle the reverse `ImageSampler` into separate `Image` and `Sampler` case at all.

I'm still hoping naga eventually supports this...

**Comment by tombh** [comment]
Just to add to @glfmn's [comment](https://github.com/gfx-rs/wgpu/issues/4342#issuecomment-1780094604), I also got stuck on how to call a combined sampled texture. It's not `texture(u_sampler, coords)` it's `texture(sampler2D(u_texture, u_sampler), coords)`.

**Comment by davnotdev** [comment]
Should have shared much earlier, but if this is of use to anyone, I also made a crate (with cli + ffi) to address this called [`spirv-webgpu-transform`](https://github.com/davnotdev/spirv-webgpu-transform), addressing this issue among others. I've been using this for a while to build Godot's WebGPU driver and it works wonders.

But hey, I hope someone gets to this feature at some point!
